libphal/phal_dump.C

```diff
@@ -0,0 +1,299 @@
+		}
+
+		if (pdbg_target_probe(target) != PDBG_TARGET_ENABLED) {
+			break;
+		}
+		proc = target;
+	}
+	if (proc == NULL) {
+		log(level::ERROR,
+		    "No Proc target found to execute the dump for failing "
+		    "unit=%d",
+		    failingUnit);
+		throw sbeError_t(exception::PDBG_TARGET_NOT_OPERATIONAL);
+	}
+	return proc;
+}
+
+/**
+ * @brief Initialize the PDBG and SBE to start collection
+ * @param[in] failingUnit Position of the proc containing the failed SBE
+ *
+ * Exceptions: PDBG_INIT_FAIL for any pdbg init related failure.
+ *             PDBG_TARGET_NOT_OPERATIONAL for invalid failing unit
+ *             HWP_EXECUTION_FAILURE if the extract rc procedure is failing
+ */
+struct pdbg_target* preCollection(const uint32_t failingUnit)
+{
+	// Initialize PDBG with KERNEL backend
+	pdbg::init(PDBG_BACKEND_KERNEL);
+
+	// FSI and PIB targets for executing HWP
+	struct pdbg_target *fsi, *pib;
+	char path[16];
+
+	// Find the proc target from the failing unit id
+	struct pdbg_target* proc = getProcFromFailingId(failingUnit);
+
+	// Probe FSI for HWP execution
+	sprintf(path, "/proc%d/fsi", pdbg_target_index(proc));
+	fsi = pdbg_target_from_path(NULL, path);
+	if (fsi == nullptr) {
+		log(level::ERROR, "Failed to get FSI target for(%s)",
+		    pdbg_target_path(proc));
+		throw dumpError_t(exception::PDBG_TARGET_NOT_OPERATIONAL);
+	}
+
+	// Probe PIB for HWP execution
+	sprintf(path, "/proc%d/pib", pdbg_target_index(proc));
+	pib = pdbg_target_from_path(NULL, path);
+	if (pib == nullptr) {
+		log(level::ERROR, "Failed to get PIB target for(%s)",
+		    pdbg_target_path(proc));
+		throw dumpError_t(exception::PDBG_TARGET_NOT_OPERATIONAL);
+	}
+
+	// Execute SBE extract rc to set up sdb bit for pibmem dump to work
+	// TODO Add error handling or revisit procedure later
+	fapi2::ReturnCode fapiRc;
+	P10_EXTRACT_SBE_RC::RETURN_ACTION recovAction;
+	fapiRc = p10_extract_sbe_rc(proc, recovAction, true);
+	if (fapiRc != fapi2::FAPI2_RC_SUCCESS) {
+		log(level::ERROR,
+		    "Failed in p10_extract_sbe_rc for proc=%s, rc=0x%08X",
+		    pdbg_target_path(proc), fapiRc);
+		throw dumpError_t(exception::HWP_EXECUTION_FAILURE);
+	}
+
+	log(level::INFO, "SBE RC=0x%08X", recovAction);
+	return proc;
+}
+
+/**
+ * @brief Write the dump file to the path
+ * @param[in] data Pointer to the data buffer
+ * @param[in] len Length of the buffer
+ * @param[in] dumpPath Path of the file
+ *
+ * Throws FILE_OPERATION_FAILURE in the case of error
+ */
+void writeDumpFile(char* data, size_t len, std::filesystem::path& dumpPath)
+{
+	std::ofstream outfile{dumpPath, std::ios::out | std::ios::binary};
+	if (!outfile.good()) {
+		int err = errno;
+		log(level::ERROR,
+		    "Failed to open the dump file for writing err=%d", err);
+		throw dumpError_t(exception::FILE_OPERATION_FAILURE);
+	}
+	outfile.exceptions(std::ifstream::failbit | std::ifstream::badbit |
+			   std::ifstream::eofbit);
+	try {
+		outfile.write(data, len);
+	} catch (std::ofstream::failure& oe) {
```

Comment:
can we use generic errors like these unlesss there is a reason, meaning is better conveyed with those exceptions.

http://fstone07p1.aus.stglabs.ibm.com:8080/source/xref/openbmc/phosphor-logging/extensions/openpower-pels/repository.cpp?r=afb1b46f#197

In the caller you could do this
try
{
    writeDumpFile(attr->second.path, func);
}
catch (std::exception& e)
{
    log<level::ERR>("Unable to update PEL host transmission state",
            entry("PATH=%s", attr->second.path.c_str()),
            entry("ERROR=%s", e.what()));
}

Reply:
I see here the dumpError thrown is simply logged in the caller.

suggestion, you can decide